File: ruby/udemy_class/linked_list/detect_cycle.rb
class Node

  attr_accessor :next_node

  def initialize(value, next_node = nil)
    @value = value
    @next_node = next_node
  end
end


# Return true if it contains a cycle
def detect_cycle_with_hash(head)
  seen_nodes = {}

  current_node = head

  while current_node != nil
    if seen_nodes[current_node.object_id]
      return true
    end

    seen_nodes[current_node.object_id] = true
    current_node = current_node.next_node
  end

  return false
end

a = Node.new(1)
b = Node.new(2)
c = Node.new(3)

a.next_node = b
b.next_node = c

puts(detect_cycle_with_hash(a))

c.next_node = a
puts(detect_cycle_with_hash(a))

def detect_cycle_two_pointer(head)
  current_node = head
  pointer = current_node

  while pointer && pointer.next_node
    current_node = current_node.next_node
    pointer = pointer.next_node.next_node

    if current_node == pointer
      return true
    end
  end

  return false
end

a = Node.new(1)
b = Node.new(2)
c = Node.new(3)

a.next_node = b
b.next_node = c

puts(detect_cycle_two_pointer(a))

c.next_node = a
puts(detect_cycle_two_pointer(a))

File: python/udemy_class/linked_list/detect_cycle.py
class Node(object):

    def __init__(self, value, next_node = None):
        self.value = value
        self.next_node = next_node


def detect_cycle_with_dict(head):
    nodes_seen = {}

    current_node = head

    while current_node != None:
        if current_node in nodes_seen:
            return True

        nodes_seen[current_node] = True
        current_node = current_node.next_node

    return False

a = Node(1)
b = Node(2)
c = Node(3)

a.next_node = b
b.next_node = c

print(detect_cycle_with_dict(a))

c.next_node = a
print(detect_cycle_with_dict(a))

def detect_cycle_two_pointer(head):
    pointer_one = head
    pointer_two = head


    # We dont need to check if pointer one is null
    # because pointer two will always be head.
    # If pointer two reaches null then we know there is no cycle
    while pointer_two != None and pointer_two.next_node != None:
        pointer_one = pointer_one.next_node
        pointer_two = pointer_two.next_node.next_node

        if pointer_one == pointer_two:
            return True

    return False

a = Node(1)
b = Node(2)
c = Node(3)

a.next_node = b
b.next_node = c

print(detect_cycle_two_pointer(a))

c.next_node = a
print(detect_cycle_two_pointer(a))
